File: bMode/bMode.cpp
/**
 * B-mode Ultrasound Reconstruction
 * Converted from MATLAB script.
 *
 * Dependencies:
 *   - libmatio  (reading .mat files):  sudo apt install libmatio-dev
 *   - FFTW3     (Hilbert transform):   sudo apt install libfftw3-dev
 *   - matplotlib-cpp or gnuplot (optional, for plotting)
 *
 * Compile:
 *   g++ -O2 -std=c++17 bmode_scan.cpp -o bmode_scan -lmatio -lfftw3
 *
 * Usage:
 *   ./bmode_scan
 *   (edit the Settings section below as needed)
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>
#include <stdexcept>
#include <glob.h>       // POSIX glob for file pattern matching
#include <matio.h>      // libmatio for .mat file reading
#include <fftw3.h>      // FFTW for FFT-based Hilbert transform

// ============================================================
// Settings (edit if needed)
// ============================================================
static const char*  FILE_PATTERN  = "20260225_*.mat";
static const double C             = 1480.0;   // speed of sound in water (m/s)
static const double DYN_RANGE_DB  = 50.0;     // display dynamic range (dB)
static const bool   USE_TGC       = true;     // time gain compensation on/off
static const double TGC_ALPHA     = 1.5;      // TGC strength (0.5 – 3)

// ============================================================
// Helper: glob file list
// ============================================================
std::vector<std::string> globFiles(const std::string& pattern)
{
    glob_t result{};
    std::vector<std::string> files;
    if (glob(pattern.c_str(), GLOB_TILDE, nullptr, &result) == 0) {
        for (size_t i = 0; i < result.gl_pathc; ++i)
            files.emplace_back(result.gl_pathv[i]);
    }
    globfree(&result);
    std::sort(files.begin(), files.end());
    return files;
}

// ============================================================
// Helper: load a scalar double field from a mat_t
// ============================================================
double loadScalar(mat_t* mat, const char* name, const std::string& fname)
{
    matvar_t* v = Mat_VarRead(mat, name);
    if (!v) throw std::runtime_error("Missing field '" + std::string(name) + "' in " + fname);
    double val = *static_cast<double*>(v->data);
    Mat_VarFree(v);
    return val;
}

// ============================================================
// Helper: load a 1-D double array field from a mat_t
// ============================================================
std::vector<double> loadArray(mat_t* mat, const char* name, const std::string& fname)
{
    matvar_t* v = Mat_VarRead(mat, name);
    if (!v) throw std::runtime_error("Missing field '" + std::string(name) + "' in " + fname);
    size_t n = v->dims[0] * v->dims[1];
    std::vector<double> out(n);
    double* ptr = static_cast<double*>(v->data);
    std::copy(ptr, ptr + n, out.begin());
    Mat_VarFree(v);
    return out;
}

// ============================================================
// Hilbert envelope via FFTW  (in-place, column by column)
// Returns |analytic signal| for each column of B (N x K)
// ============================================================
std::vector<double> hilbertEnvelope(const std::vector<double>& B, size_t N, size_t K)
{
    std::vector<double> env(N * K);

    // Allocate FFTW buffers once
    fftw_complex* freq = fftw_alloc_complex(N);
    double*       buf  = fftw_alloc_real(N);

    fftw_plan planFwd = fftw_plan_dft_r2c_1d((int)N, buf, freq, FFTW_ESTIMATE);
    fftw_plan planInv = fftw_plan_dft_c2r_1d((int)N, freq, buf, FFTW_ESTIMATE);

    for (size_t k = 0; k < K; ++k) {
        // Copy column into buf
        for (size_t n = 0; n < N; ++n) buf[n] = B[n * K + k];

        // Forward FFT
        fftw_execute(planFwd);

        // Apply Hilbert weighting in frequency domain
        //   DC (0) and Nyquist stay unchanged
        //   Positive freqs [1 .. N/2-1] multiplied by 2
        //   Negative freqs [N/2+1 .. N-1] set to 0
        size_t Nhalf = N / 2;
        for (size_t i = 1; i < Nhalf; ++i) {
            freq[i][0] *= 2.0;
            freq[i][1] *= 2.0;
        }
        for (size_t i = Nhalf + 1; i < N; ++i) {
            freq[i][0] = 0.0;
            freq[i][1] = 0.0;
        }

        // Inverse FFT
        fftw_execute(planInv);

        // Compute envelope magnitude, normalise by N (FFTW convention)
        // The imaginary part of the analytic signal is in buf (real part of IFFT output)
        // We need both real and imaginary; reconstruct:
        //   real(analytic) = original signal
        //   imag(analytic) = IFFT result / N  (from the doubled positive freqs trick)
        for (size_t n = 0; n < N; ++n) {
            double re = B[n * K + k];
            double im = buf[n] / (double)N;
            env[n * K + k] = std::sqrt(re * re + im * im);
        }
    }

    fftw_destroy_plan(planFwd);
    fftw_destroy_plan(planInv);
    fftw_free(freq);
    fftw_free(buf);

    return env;
}

// ============================================================
// Save result as a simple CSV for external plotting
// ============================================================
void saveCSV(const std::string& path,
             const std::vector<double>& env_db,
             const std::vector<double>& depth_mm,
             size_t N, size_t K)
{
    std::ofstream f(path);
    if (!f) throw std::runtime_error("Cannot open " + path + " for writing");

    // Header: depth_mm, col1, col2, ...
    f << "depth_mm";
    for (size_t k = 0; k < K; ++k) f << ",col" << (k + 1);
    f << "\n";

    for (size_t n = 0; n < N; ++n) {
        f << depth_mm[n];
        for (size_t k = 0; k < K; ++k)
            f << "," << env_db[n * K + k];
        f << "\n";
    }
}

// ============================================================
// Main
// ============================================================
int main()
{
    // --- Load file list ---
    std::vector<std::string> files = globFiles(FILE_PATTERN);
    if (files.empty()) {
        std::cerr << "No files found matching: " << FILE_PATTERN << "\n";
        return 1;
    }
    size_t K = files.size();
    std::cout << "Found " << K << " file(s).\n";

    // --- Load first file for timing ---
    mat_t* mat0 = Mat_Open(files[0].c_str(), MAT_ACC_RDONLY);
    if (!mat0) throw std::runtime_error("Cannot open " + files[0]);

    std::vector<double> A0 = loadArray(mat0, "A", files[0]);
    double Tstart    = loadScalar(mat0, "Tstart",    files[0]);
    double Tinterval = loadScalar(mat0, "Tinterval", files[0]);
    Mat_Close(mat0);

    size_t N = A0.size();

    // Build time vector
    std::vector<double> t(N);
    for (size_t i = 0; i < N; ++i)
        t[i] = Tstart + (double)i * Tinterval;

    // --- Load all A-scans into B (row-major: B[n*K + k]) ---
    std::vector<double> B(N * K, 0.0);

    for (size_t k = 0; k < K; ++k) {
        mat_t* mat = Mat_Open(files[k].c_str(), MAT_ACC_RDONLY);
        if (!mat) { std::cerr << "Cannot open " << files[k] << ", skipping.\n"; continue; }

        std::vector<double> a = loadArray(mat, "A", files[k]);
        Mat_Close(mat);

        if (a.size() != N) {
            std::cerr << "Warning: length mismatch in " << files[k]
                      << " (got " << a.size() << ", expected " << N << "). Trunc/pad applied.\n";
            a.resize(N, 0.0);
        }
        for (size_t n = 0; n < N; ++n)
            B[n * K + k] = a[n];
    }

    // --- Crop pre-trigger (t < 0) ---
    size_t startIdx = 0;
    while (startIdx < N && t[startIdx] < 0.0) ++startIdx;

    N -= startIdx;
    t.erase(t.begin(), t.begin() + (long)startIdx);
    std::vector<double> Bcrop(N * K);
    for (size_t n = 0; n < N; ++n)
        for (size_t k = 0; k < K; ++k)
            Bcrop[n * K + k] = B[(n + startIdx) * K + k];
    B = std::move(Bcrop);

    // --- Envelope detection (Hilbert) ---
    std::vector<double> env = hilbertEnvelope(B, N, K);

    // Normalize
    double maxEnv = *std::max_element(env.begin(), env.end());
    for (auto& v : env) v /= (maxEnv + 1e-300);

    // --- Optional TGC ---
    if (USE_TGC) {
        double tMax = t.back();
        for (size_t n = 0; n < N; ++n) {
            double g = std::exp(TGC_ALPHA * (t[n] / tMax));
            for (size_t k = 0; k < K; ++k)
                env[n * K + k] *= g;
        }
        maxEnv = *std::max_element(env.begin(), env.end());
        for (auto& v : env) v /= (maxEnv + 1e-300);
    }

    // --- Log compression (dB) ---
    std::vector<double> env_db(N * K);
    for (size_t i = 0; i < N * K; ++i)
        env_db[i] = 20.0 * std::log10(env[i] + 1e-12);

    // --- Convert time -> depth (mm) ---
    std::vector<double> depth_mm(N);
    for (size_t n = 0; n < N; ++n)
        depth_mm[n] = (C * t[n] / 2.0) * 1000.0;

    // --- Save to CSV (plot with Python/MATLAB/gnuplot) ---
    const std::string outCSV = "bmode_output.csv";
    saveCSV(outCSV, env_db, depth_mm, N, K);
    std::cout << "Saved B-mode data to " << outCSV << "\n";
    std::cout << "Depth range: " << depth_mm.front() << " – " << depth_mm.back() << " mm\n";
    std::cout << "Dynamic range: " << DYN_RANGE_DB << " dB\n";

    // --- Optional: quick plot with gnuplot (if available) ---
    // Uncomment the lines below if gnuplot is installed:
    // std::string cmd = "gnuplot -e \""
    //     "set terminal png size 1200,600; set output 'bmode.png';"
    //     "set yrange [" + std::to_string(depth_mm.back()) + ":" + std::to_string(depth_mm.front()) + "];"
    //     "set cbrange [" + std::to_string(-DYN_RANGE_DB) + ":0];"
    //     "set palette gray; plot 'bmode_output.csv' matrix with image\"";
    // system(cmd.c_str());

    return 0;
}

File: bMode/bMode.py
"""
B-mode Ultrasound Reconstruction
Input: one CSV per A-scan capture, with columns:
    Time (ms)  |  Channel A (V)
    (first two rows are header + unit label, data starts row 3)

Dependencies:
    pip install numpy scipy matplotlib pandas
"""

import glob
import numpy as np
import pandas as pd
from scipy.signal import hilbert
import matplotlib.pyplot as plt

# =========================
# Settings (edit if needed)
# =========================
pattern     = r'C:\Users\omarS\Documents\Capstone\Data2/dup\avg_*.csv'   # your file pattern
c           = 1480               # speed of sound in water (m/s)
dynRange_dB = 50                 # display dynamic range (e.g., 40-70)
useTGC      = True               # time gain compensation on/off
tgc_alpha   = 2                # strength of TGC (try 0.5 to 3)

# =========================
# Load file list
# =========================
files = sorted(glob.glob(pattern))
if not files:
    raise FileNotFoundError(f'No files found matching: {pattern}')

# =========================
# Helper: load one CSV A-scan
# CSV format:
#   Row 0: "Time"  "Channel A"         <- column names
#   Row 1: "(ms)"  "(V)"               <- units  (skipped)
#   Row 2+: numeric data
# =========================
def load_csv(path: str):
    # Read with header on row 0, then drop the units row (index 0 after read)
    df = pd.read_csv(path, header=0, skiprows=[1])
    df.columns = [c.strip() for c in df.columns]   # tidy whitespace

    # Accept various column name spellings
    time_col = next((c for c in df.columns if 'time' in c.lower()), None)
    volt_col = next((c for c in df.columns if 'channel' in c.lower() or
                     'ch' in c.lower() or 'volt' in c.lower() or c.strip() == 'A'), None)

    if time_col is None or volt_col is None:
        raise ValueError(f"Cannot identify Time/Voltage columns in {path}.\n"
                         f"Found columns: {list(df.columns)}")

    t_ms = pd.to_numeric(df[time_col], errors='coerce').to_numpy(dtype=np.float64)
    v    = pd.to_numeric(df[volt_col], errors='coerce').to_numpy(dtype=np.float64)

    # Drop any NaN rows
    mask = np.isfinite(t_ms) & np.isfinite(v)
    return t_ms[mask], v[mask]

# =========================
# Load first file for timing reference
# =========================
t_ms_ref, a_ref = load_csv(files[0])
N = len(t_ms_ref)
t = t_ms_ref * 1e-3   # convert ms -> seconds

K = len(files)
B = np.zeros((N, K), dtype=np.float32)

# =========================
# Load all A-scans
# =========================
for k, fname in enumerate(files):
    t_ms_k, a = load_csv(fname)
    a = a.astype(np.float32)

    if len(a) != N:
        print(f'Warning: length mismatch in {fname} '
              f'(got {len(a)}, expected {N}). Trunc/pad applied.')
        aa = np.zeros(N, dtype=np.float32)
        nmin = min(N, len(a))
        aa[:nmin] = a[:nmin]
        a = aa

    B[:, k] = a

# =========================
# Crop pre-trigger (t < 0)
# =========================
valid = t >= 0
t = t[valid]
B = B[valid, :]

# =========================
# Envelope detection (Hilbert)
# =========================
B = B - np.mean(B, axis=0)
env = np.abs(hilbert(B.astype(np.float64), axis=0))   # N x K
env = env / (env.max() + np.finfo(float).eps)


# =========================
# Optional TGC (boost deeper echoes)
# =========================
if useTGC:
    g   = np.exp(tgc_alpha * (t / t.max()))   # shape (N,)
    env = env * g[:, np.newaxis]
    env = env / (env.max() + np.finfo(float).eps)

# =========================
# Log compression (dB)
# =========================
env_db = 20 * np.log10(env + 1e-12)

# =========================
# Convert time -> depth (mm)
# =========================
depth_mm = (c * t / 2) * 1000

# =========================
# Display B-mode style image
# =========================
fig, ax = plt.subplots(figsize=(10, 6))
im = ax.imshow(
    env_db,
    aspect='auto',
    cmap='gray',
    vmin=-dynRange_dB, vmax=0,
    extent=[1, K, depth_mm[-1], depth_mm[0]],
    origin='upper'
)
plt.colorbar(im, ax=ax, label='dB')
ax.set_xlabel('Capture Index')
ax.set_ylabel('Depth (mm)')
ax.set_title(f'B-mode (Envelope + Log) | Water c={c} m/s | DR={dynRange_dB} dB')
plt.tight_layout()
plt.show()

